fit/curveNd_test.go

```diff
@@ -0,0 +1,189 @@
+// Copyright ©2017 The go-hep Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package fit_test
+
+import (
+	"bufio"
+	//"image/color"
+	"log"
+	"math/rand"
+	"os"
+	"strconv"
+	"strings"
+	"testing"
+
+	"go-hep.org/x/hep/fit"
+	"go-hep.org/x/hep/hbook"
+	"go-hep.org/x/hep/hplot"
+	"gonum.org/v1/gonum/floats"
+	"gonum.org/v1/gonum/optimize"
+	"gonum.org/v1/plot/plotter"
+	"gonum.org/v1/plot/vg"
+)
+
+func TestCurve2D(t *testing.T) {
+	ExampleCurve2D_plane()
+}
+
+func ExampleCurve2D_plane() {
+	var (
+		m1    = 0.3
+		m2    = 0.1
+		c     = 0.2
+		ps    = []float64{m1, m2, c}
+		n0    = 30
+		n1    = 30
+		x0min = -1.
+		x0max = 1.
+		x1min = -1.
+		x1max = 1.
+	)
+
+	plane := func(x, ps []float64) float64 {
+		return ps[0]*x[0] + ps[1]*x[1] + ps[2]
+	}
+
+	xdata, ydata := genData2D(n0, n1, plane, ps, x0min, x0max, x1min, x1max)
+
+	res, err := fit.CurveND(
+		fit.FuncND{
+			F: func(x []float64, ps []float64) float64 {
+				return plane(x, ps)
+			},
+			X: xdata,
+			Y: ydata,
+			N: 3,
+		},
+		nil, &optimize.NelderMead{},
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if err := res.Status.Err(); err != nil {
+		log.Fatal(err)
+	}
+	if got, want := res.X, []float64{m1, m2, c}; !floats.EqualApprox(got, want, 0.1) {
+		log.Fatalf("got= %v\nwant=%v\n", got, want)
+	}
+
+	{
+		p := hplot.New()
+		p.X.Label.Text = "x"
+		p.Y.Label.Text = "y"
+		p.Y.Min = x1min
+		p.Y.Max = x1max
+		p.X.Min = x0min
+		p.X.Max = x0max
+
+		s := hbook.NewH2D(n0, x0min, x0max, n1, x1min, x1max)
+
+		for i := range xdata {
+			s.Fill(xdata[i][0], xdata[i][1], ydata[i])
+		}
+
+		p.Add(hplot.NewH2D(s, nil))
+		p.Add(plotter.NewGrid())
+
+		err := p.Save(20*vg.Centimeter, -1, "testdata/2DPlane-plot.png")
```

Comment:
```suggestion
		err := p.Save(20*vg.Centimeter, -1, "testdata/2dplane-plot.png")
```
(easier to have all file names as lower case, for filesystems that aren't case sensitive.)

please also add that file to the PR (so we can include it inside the "README.md" file).
it would be great to also add the example to that README, using the "embedmd" tool:
```
![2d-example](https://github.com/go-hep/hep/raw/master/fit/testdata/2dplane-plot.png)

[embedmd]:# (curvend_test.go go /func ExampleCurve2D_plane/ /\n}/)
```

Reply:
Done. I modified the go code embedding in this file, ` ```go` causes embedmd to expect a package to be embedded, with the space ` ``` go` it then expects a function.

Reply:
interesting...
how did you invoke `embedmd`?
(I tried to reproduce with the latest `v1` version (I haven't tried `v2`))

theoretically, I've wired up `embedmd` with `go generate` in the `hep/fit` directory, so (logically) you'd only have to:
```
$> cd hep/fit
$> go generate -v
```

Reply:
(but ok. we can leave this like that. it's not super important)

Reply:
Cool (I'm not a great source for info on embedmd, I only noticed the above because my IDE showed it as an error)